File: algorithms/python/数组中的第K个最大元素/数组中的第K个最大元素.py
# Source : https://leetcode.cn/problems/kth-largest-element-in-an-array/description/
# Author : Yanan Wang
# Date   : 2023-08-22

# 本题采用快速选择算法（时间复杂度为O(N)，证明自《算法导论》）：基于快速排序（分区分三块）
# 当快速排序完成一次时，nums会分成三个区域：
#      -----------------------
#      ^          ^          ^ 
#     <key      ==key       >key  
#      a个        b个        c个
# 本题求的是第K个最大元素，那么就根据这三个分区的数量来选择下次递归的范围
# 当 c >= k时：去右边区域寻找第k个最大元素
# 当 b + c >= k时：直接返回key即可，因为此时第k个元素肯定落在中间区域
# 当上面两个条件都不成立时：说明第k个元素落在左边区域，那么就去左边区域寻找第k-b-c个最大元素

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def quickSelect(nums, low, high, k):
            if low == high:
                return nums[low]
            key = nums[random.randint(low, high)]
            left, right = low, high
            i = low
            while i <= right:
                if nums[i] < key:
                    nums[left], nums[i] = nums[i], nums[left]
                    left += 1
                    i += 1
                elif nums[i] == key:
                    i += 1
                else:
                    nums[i], nums[right] = nums[right], nums[i]
                    right -= 1
            # [low, left)  [left, right]  (right, high]
            # left - low   right-left+1    high - right  
            leftNums = left - low
            midNums = right - left + 1
            rightNums = high - right
            if rightNums >= k:
                return quickSelect(nums, right + 1, high, k)
            elif midNums + rightNums >= k:
                return key
            else:
                return quickSelect(nums, low, left - 1, k - midNums - rightNums)


        return quickSelect(nums, 0, len(nums) - 1, k)

File: algorithms/cpp/数组中的第K个最大元素/数组中的第K个最大元素.cpp
// Source : https://leetcode.cn/problems/kth-largest-element-in-an-array/description/
// Author : Yanan Wang
// Date   : 2023-11-25

class Solution {
public:
    int QuickSelect(vector<int> &nums, int begin, int end, int k) {
        if (begin == end) {
            return nums[begin];
        }
        int key = nums[begin + (rand() % (end - begin + 1))];
        int i = begin;
        int left = begin - 1;
        int right = end + 1;
        while (i < right) {
            if (nums[i] > key) {
                swap(nums[++left], nums[i++]);
            } else if (nums[i] == key) {
                ++i;
            } else {
                swap(nums[i], nums[--right]);
            }
        }
        int left_nums = left - begin + 1;
        int mid_nums = right - left - 1;
        if (left_nums >= k) {
            return QuickSelect(nums, begin, left, k);
        } else if (left_nums + mid_nums >= k) {
            return key;
        } else {
            return QuickSelect(nums, right, end, k - left_nums - mid_nums);
        }
    }
    int findKthLargest(vector<int>& nums, int k) {
        srand(time(nullptr));
        return QuickSelect(nums, 0, nums.size() - 1, k);
    }
};